VitruvioHost/Plugins/Vitruvio/Source/VitruvioEditor/Private/VitruvioCooker.cpp

```diff
@@ -0,0 +1,332 @@
+		PRTGenerateCallsTasks.EnterProgressFrame(TotalGenerateCalls - CurrentNumGenerateCalls);
+		TotalGenerateCalls = CurrentNumGenerateCalls;
+	}
+}
+
+UTexture2D* SaveTexture(UTexture2D* Original, const FString& Path, FTextureCache& TextureCache)
+{
+	if (TextureCache.Contains(Original))
+	{
+		return TextureCache[Original];
+	}
+
+	FString TexturePath = FPaths::Combine(Path, TEXT("Textures"), Original->GetName());
+	UPackage* TexturePackage = CreatePackage(*TexturePath);
+	UTexture2D* NewTexture = NewObject<UTexture2D>(TexturePackage, *Original->GetName(), RF_Public | RF_Standalone);
+
+	NewTexture->PlatformData = new FTexturePlatformData();
+	NewTexture->PlatformData->SizeX = Original->PlatformData->SizeX;
+	NewTexture->PlatformData->SizeY = Original->PlatformData->SizeY;
+	NewTexture->PlatformData->PixelFormat = Original->PlatformData->PixelFormat;
+	NewTexture->CompressionSettings = Original->CompressionSettings;
+	NewTexture->SRGB = Original->SRGB;
+
+	// Allocate first mipmap and upload the pixel data
+	FTexture2DMipMap* Mip = new FTexture2DMipMap();
+	FTexture2DMipMap& OriginalMip = Original->PlatformData->Mips[0];
+
+	NewTexture->PlatformData->Mips.Add(Mip);
+
+	Mip->SizeX = OriginalMip.SizeX;
+	Mip->SizeY = OriginalMip.SizeY;
+
+	const uint8* SourcePixels = static_cast<const uint8*>(OriginalMip.BulkData.LockReadOnly());
+	Mip->BulkData.Lock(LOCK_READ_WRITE);
+
+	void* TextureData = Mip->BulkData.Realloc(CalculateImageBytes(Mip->SizeX, Mip->SizeY, 0, NewTexture->PlatformData->PixelFormat));
+	FMemory::Memcpy(TextureData, SourcePixels, OriginalMip.BulkData.GetBulkDataSize());
+	Mip->BulkData.Unlock();
+
+	NewTexture->Source.Init(Original->PlatformData->SizeX, Original->PlatformData->SizeY, 1, 1, ETextureSourceFormat::TSF_BGRA8, SourcePixels);
+	OriginalMip.BulkData.Unlock();
+
+	NewTexture->PostEditChange();
+	TexturePackage->MarkPackageDirty();
+	FAssetRegistryModule::AssetCreated(NewTexture);
+
+	TextureCache.Add(Original, NewTexture);
+
+	return NewTexture;
+}
+
+UMaterialInstanceConstant* SaveMaterial(UMaterialInstanceDynamic* Material, const FString& Path, FMaterialCache& MaterialCache,
+										FTextureCache& TextureCache)
+{
+	if (MaterialCache.Contains(Material))
+	{
+		return MaterialCache[Material];
+	}
+
+	const FString MaterialName = Material->GetName();
+	const FString MaterialPackagePath = FPaths::Combine(Path, TEXT("Materials"), MaterialName);
+	UPackage* MaterialPackage = CreatePackage(*MaterialPackagePath);
+
+	UMaterialInstanceConstantFactoryNew* MaterialFactory = NewObject<UMaterialInstanceConstantFactoryNew>();
+	MaterialFactory->InitialParent = Material->Parent;
+
+	UMaterialInstanceConstant* NewMaterial = Cast<UMaterialInstanceConstant>(MaterialFactory->FactoryCreateNew(
+		UMaterialInstanceConstant::StaticClass(), MaterialPackage, *MaterialName, RF_Public | RF_Standalone, nullptr, GWarn));
+	FAssetRegistryModule::AssetCreated(NewMaterial);

```

Comment:
Is there a reason why this call is placed here and not at the end of the function? (for the texture/mesh its after `PostEditChange/MarkPackageDirty`